### PR TITLE
Add random delay to ArchiveExecutionTask

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
1. I made the archival queue a scheduled queue instead of a timer queue.
2. I added a random delay to each task we produce

**Why?**
I did this to prevent us from having a ton of archive execution tasks arrive concurrently if many workflows close simultaneously.

**How did you test it?**
I added unit tests that verify edge cases like setting the dynamic config upper bound to 0.

**Potential risks**
Does not run via any production entrypoints. 

**Is hotfix candidate?**
No.

### CODE REVIEW

File: service/history/workflow/task_generator.go [hunk 1]
```diff
@@ -218,6 +236,34 @@ func (r *TaskGeneratorImpl) GenerateWorkflowCloseTasks(
 	return nil
 }
 
+func (r *TaskGeneratorImpl) getArchivalTaskDelay(namespaceEntry *namespace.Namespace, retention time.Duration) time.Duration {
```
Comment (yycptt) [inline service/history/workflow/task_generator.go:239]: Please check and reuse `backoff.JitDuration`.

Comment (MichaelSnowden) [reply service/history/workflow/task_generator.go]: Fixed, thanks for pointing that out.

**Review by yycptt — COMMENTED** [review]
Archival queue is an immediate queue, there's no delay by definition & task visibility timestamp for immediate tasks (which is only for metrics purpose) will be overwritten to time.Now() by shard context.

File: service/history/workflow/task_generator.go [hunk 2]
```diff
@@ -102,6 +103,7 @@ type (
 		namespaceRegistry namespace.Registry
 		mutableState      MutableState
 		config            *configs.Config
+		logger            log.Logger
```
Comment (yux0) [inline service/history/workflow/task_generator.go:106]: why do it need logger here?

File: service/history/configs/config.go [hunk 3]
```diff
@@ -303,6 +304,8 @@ type Config struct {
 	ArchivalProcessorMaxPollIntervalJitterCoefficient   dynamicconfig.FloatPropertyFn
 	ArchivalProcessorUpdateAckInterval                  dynamicconfig.DurationPropertyFn
 	ArchivalProcessorUpdateAckIntervalJitterCoefficient dynamicconfig.FloatPropertyFn
+	ArchivalProcessorArchiveDelay                       dynamicconfig.DurationPropertyFn
+	ArchivalProcessorArchiveDelayJitterCoefficient      dynamicconfig.FloatPropertyFn
```
Comment (yux0) [inline service/history/configs/config.go:308]: for simplicity, let's use 1 for add jitter.

File: service/history/workflow/task_generator.go [hunk 4]
```diff
@@ -198,10 +202,20 @@ func (r *TaskGeneratorImpl) GenerateWorkflowCloseTasks(
 			},
 		)
 		if r.config.DurableArchivalEnabled() {
+			delay := backoff.JitDuration(
```
Comment (yux0) [inline service/history/workflow/task_generator.go:205]: `backoff.JitDuration(delay, 1) / 2`

**Comment by yux0** [comment]
> Archival queue is an immediate queue, there's no delay by definition & task visibility timestamp for immediate tasks (which is only for metrics purpose) will be overwritten to time.Now() by shard context.

With task processing rate limiter, do we still need the jitter here? I think this is the main reason we want to change it to scheduled queue.

File: service/history/workflow/task_generator.go (hunk 2 above)
Comment (MichaelSnowden) [reply service/history/workflow/task_generator.go]: Probably from an older revision. Removed, thanks for spotting!

File: service/history/configs/config.go (hunk 3 above)
Comment (MichaelSnowden) [reply service/history/configs/config.go:308]: Got it, removed

File: service/history/configs/config.go [hunk 5]
```diff
@@ -186,10 +186,11 @@ type Config struct {
...
 	// Archival settings
-	NumArchiveSystemWorkflows dynamicconfig.IntPropertyFn
-	ArchiveRequestRPS         dynamicconfig.IntPropertyFn
-	ArchiveSignalTimeout      dynamicconfig.DurationPropertyFn
-	DurableArchivalEnabled    dynamicconfig.BoolPropertyFn
+	NumArchiveSystemWorkflows             dynamicconfig.IntPropertyFn
+	ArchiveRequestRPS                     dynamicconfig.IntPropertyFn
+	ArchiveSignalTimeout                  dynamicconfig.DurationPropertyFn
+	DurableArchivalEnabled                dynamicconfig.BoolPropertyFn
+	RandomArchiveExecutionDelayUpperBound dynamicconfig.DurationPropertyFn
```
Comment (yycptt) [inline service/history/configs/config.go:193]: Is this used? Remove?

**Review by yycptt — APPROVED** [review]

**Comment by yycptt** [comment]
> > Archival queue is an immediate queue, there's no delay by definition & task visibility timestamp for immediate tasks (which is only for metrics purpose) will be overwritten to time.Now() by shard context.
> 
> With task processing rate limiter, do we still need the jitter here? I think this is the main reason we want to change it to scheduled queue.

If we can avoid the spiky load in the first place, I think it's better than relying on the rate limiter to handle it after is occurs, which exists a possibility that that rate limiter can't handle it well (especially for tasks that needs to read history.) Plus, I don't think scheduled queue is a lot more expensive than immediate queues (especially when load is high).

File: service/history/configs/config.go (hunk 5 above)
Comment (MichaelSnowden) [reply service/history/configs/config.go]: Removed, thanks!
